cranelift/codegen/src/isa/riscv64/inst/mod.rs

```diff
@@ -640,6 +640,32 @@ fn riscv64_get_operands<F: Fn(VReg) -> VReg>(inst: &Inst, collector: &mut Operan
             // gen_prologue is called at emit stage.
             // no need let reg alloc know.
         }
+        &Inst::VecAluRRRImm5 {
+            op,
+            vd,
+            vd_src,
+            vs2,
+            ref mask,
+            ..
+        } => {
+            debug_assert_eq!(vd_src.class(), RegClass::Vector);
+            debug_assert_eq!(vd.to_reg().class(), RegClass::Vector);
+            debug_assert_eq!(vs2.class(), RegClass::Vector);
+
+            collector.reg_use(vs2);
+
+            // If the operation forbids source/destination overlap. In other
+            // instructions we use `early_def`, but that doesen't seem available
+            // for reuse. Use `late_use` instead which seems to do the right thing.
+            if op.forbids_src_dst_overlaps() {
+                collector.reg_late_use(vd_src);
+            } else {
+                collector.reg_use(vd_src);
+            }
+
+            collector.reg_reuse_def(vd, 1); // `vd` == `vd_src`.
```

Comment:
The `vslideup` instruction both modifies the destination register and also has the requirement that none of the input registers must be the same as the destination register.

I've used `reg_late_use` and `reg_reuse_def` to model this, but this seems wrong, and it also always emits a move before the instruction even when the register is otherwise unused, so there's probably a better way of expressing this which I'm not entirely sure how

Reply:
I'm not sure I understand the intended semantics here: src and dst (`vd_src` and `vd`, rather) must not overlap, but `vd` is also a reuse of `vd_src`? Or is the non-overlap constraint between `vs2` and `vd`?

Reply:
Oops, I could have written that better. `vd` must be the same as `vd_src` (since they are encoded in the same field), and none of the other inputs (`vs2` and `vm`) must be the same register as `vd`.

`vm` (The mask register) is slightly hidden here since it's not always applicable, but when it is, it's a `fixed_use(v0)`

Reply:
Ah, OK. The right way to encode that is probably to use a late-use on `vs2` (and on `vm` when present), rather than on `vd_src`, I think.

Reply:
👍 I'm not entirely sure I did the right thing. I also couldn't find something along the lines of `reg_fixed_late_use` for `vm`, does regalloc support encoding multiple constraints by calling `reg_fixed_use` and `reg_late_use` on the same `VReg`?